### PR TITLE
Memory manager and Object Logo fixes

### PR DESCRIPTION
**Pull request by pahihu**
I have fixed the memory manager, and fixed some parts of Object Logo.

Added one new primitive, which returns the current Unix time (TIME).
In the tests/ directory, I've added some tests as well.

One of the memory manager fixes causes segfault in your 6.1 master branch,
this triggered the memory manager fix.

### CODE REVIEW

**Comment by jrincayc** [comment]
@pahihu Thank you, I will look at this pull this week.

@brianharvey You might want to look at this as well.

**Comment by jrincayc** [comment]
With commit f447096 
```
../logo UnitTests.lg 
? RunAll
--------------------------
Berkeley Logo Unit Tests
--------------------------
Test Suites: 7

[ Arithmetic Tests ]
----------------------------------------------------------
Tests.Arithmetic.Add...                      Ok
Tests.Arithmetic.Sum...                      Ok
Tests.Arithmetic.Subtract...                 Ok
Tests.Arithmetic.Difference...               Ok
Tests.Arithmetic.Minus...                    Ok
Tests.Arithmetic.Negative...                 Ok
Tests.Arithmetic.Multiply...                 Ok
Tests.Arithmetic.Product...                  Ok
Tests.Arithmetic.Divide...                   Ok
Tests.Arithmetic.Quotient...                 Ok
Tests.Arithmetic.Remainder...                Ok
Tests.Arithmetic.Modulo...                   Ok
Tests.Arithmetic.Int...                      Ok
Tests.Arithmetic.Round...                    Ok
Tests.Arithmetic.Sqrt...                     Ok
Tests.Arithmetic.Power...                    Ok
Tests.Arithmetic.Exp...                      Ok
Tests.Arithmetic.Log10...                    Ok
----------------------------------------------------------

[ Bitwise Operations ]
----------------------------------------------------------
Tests.Bitwise.BitAnd...                      Ok
Tests.Bitwise.BitOr...                       Ok
Tests.Bitwise.BitXOr...                      Ok
Tests.Bitwise.BitNot...                      Ok
Tests.Bitwise.AShift...                      Ok
Tests.Bitwise.LShift...                      Ok
----------------------------------------------------------

[ Constructors ]
----------------------------------------------------------
Tests.Constructors.Word...                   Ok
Tests.Constructors.List...                   Ok
Tests.Constructors.ListMany...               Ok
Tests.Constructors.Sentence...               Ok
Tests.Constructors.SentenceMany...           Ok
Tests.Constructors.FPut...                   Ok
Tests.Constructors.LPut...                   Ok
Tests.Constructors.Array...                  Ok
Tests.Constructors.MDArray...                Ok
Tests.Constructors.MDArrayMulti...           Ok
----------------------------------------------------------

[ Predicates ]
----------------------------------------------------------
Tests.Predicates.LessP...                    Ok
Tests.Predicates.Less?...                    Ok
Tests.Predicates.GreaterP...                 Ok
Tests.Predicates.Greater?...                 Ok
Tests.Predicates.LessEqualP...               Ok
Tests.Predicates.LessEqual?...               Ok
Tests.Predicates.GreaterEqualP...            Ok
Tests.Predicates.GreaterEqual?...            Ok
----------------------------------------------------------

[ Random Numbers ]
----------------------------------------------------------
Tests.Random.RandomNumMonadic...             Ok
Tests.Random.RandomNumDyadic...                     Error
----------------------------------------------------------

[ MemoryManager ]
----------------------------------------------------------
Tests.MemMgr.LongList...                     Ok
Tests.MemMgr.Garbage...                      Ok
----------------------------------------------------------

[ OOP Tests ]
----------------------------------------------------------
Tests.OOP.Ask...                             Ok
Tests.OOP.Exist...                           Ok
Tests.OOP.Make...                            Ok
Tests.OOP.MakeMany...                        Ok
Tests.OOP.OneOf...                           Ok
Tests.OOP.OneOfCheckVar...                   Ok
Tests.OOP.OverrideExist...                   Ok
Tests.OOP.Parents...                         Ok
Tests.OOP.TalkTo...                          Ok
Tests.OOP.Usual...                           Ok
Tests.OOP.UsualChain...                      Ok
Tests.OOP.ChildWontChangeParent...           Ok
Tests.OOP.MemberFunctionOverride...          Ok
Tests.OOP.CaseSensitiveMemberVars...         Ok
Tests.OOP.OverriddenMethods...               Ok
Tests.OOP.UsualInLogo...                     Ok
Tests.OOP.OverriddenMethods1...              Ok
Tests.OOP.OverriddenMethods2...              Ok
Tests.OOP.OverriddenMethods3...              Ok
Tests.OOP.OverriddenMethods4...              Ok
Tests.OOP.MultipleInheritanceTalker...       Ok
Tests.OOP.MultipleInheritanceShouter...      Ok
Tests.OOP.MultipleInheritanceWhisperer...    Ok
Tests.OOP.MultipleInheritanceCombination1... Ok
Tests.OOP.MultipleInheritanceCombination2... Ok
Tests.OOP.ExistChain1...                     Ok
Tests.OOP.ExistChain2...                     Ok
----------------------------------------------------------

All tests completed.
```
Impressive.

File: config.h.osx [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/* config.h.  Generated from config.h.in by configure.  */
```
Comment (jrincayc) [inline config.h.osx]: I don't think this file should be included in the repository since it should be generated by configure.

File: eval.c [hunk 2]
```diff
@@ -525,6 +529,9 @@ return(val);
     proc = procValueWithParent(fun, &parent);
     if (proc != UNDEFINED && parent != 0){ 
       usual_parent = parent;
+#ifdef DEB_USUAL_PARENT
+      fprintf(stderr,"evalProcValue: usual_parent => %p\n", usual_parent);
+#endif
```
Comment (jrincayc) [inline eval.c]: Are the DEB_USUAL_PARENT debugging statements still needed, or should they be removed?

File: eval.c [hunk 3]
```diff
@@ -656,13 +673,26 @@ return(val);
 
 fetch_cont:
     {
+#ifdef USE_GCC_DISPATCH
+#define do_label(x) &&x,
```
Comment (jrincayc) [inline eval.c]: Since this define is only used once, it would probably be better to just substitute it.

File: init.c [hunk 4]
```diff
@@ -675,7 +676,7 @@ void init(void) {
         if (message_fp != NULL) {
           printf("Changing logolib from %s to %s\n", logolib, new_logolib);
           logolib = malloc(sizeof(char)*(strlen(new_logolib)+1));
-          strcpy(logolib, new_logolib);
+          strcpy((char*) logolib, new_logolib);
```
Comment (jrincayc) [inline init.c]: Hm, why did you change logolib to const, but then cast around it here?  (Or am I misunderstanding?)

File: logo.h [hunk 5]
```diff
@@ -430,7 +430,7 @@ struct segment {
 #ifdef __RZTC__
 	struct logo_node nodes[1];
 #else
-	struct logo_node nodes[0];
+	struct logo_node nodes[1];
 #endif
```
Comment (jrincayc) [inline logo.h]: Hm, could this whole ifdef be removed?

File: makefile.osx [hunk 6]
```diff
@@ -0,0 +1,125 @@
+ifeq ($(wxlogo),yes)
```
Comment (jrincayc) [inline makefile.osx]: I don't think this file should be included in the repository since it should be generated by configure.

Side note, probably a bug should be opened about the lack of support for adding OBJECTS=1 to the makefile or logo.h in the configure script.

File: mem.c [hunk 7]
```diff
@@ -506,80 +558,89 @@ void gc(BOOLEAN no_error) {
...
-    mark(last_call);
-    mark(output_node);
-    mark(output_unode);
 
-    mark(throw_node);
-    mark(err_mesg);
-    mark(current_line);
-
-/*
+#if 1
```
Comment (jrincayc) [inline mem.c]: Can this ```#if 1``` be removed?

File: mem.c [hunk 8]
```diff
@@ -189,6 +205,35 @@ void check_valid_oldyoung(NODE *old, NODE *new) {
     }
 }
 
+void clean_oldyoungs(void) {
```
Comment (jrincayc) [inline mem.c]: Can you add a comment explaining what this function does?

File: tests/UnitTests-OOP.lg [hunk 9]
```diff
@@ -210,6 +229,154 @@ to Tests.OOP.OverriddenMethods
...
+end
+
+to Tests.OOP.OverriddenMethods4
+  OUTPUT equal? (ask :InheritC [baz4]) [fooA barB]
+end
+
+
+;;; G.Drescher: "Object-oriented Logo"
+;;; in Lawler, Yazdani: "Artifical Intelligence and Education: Learning environments and tutoring systems", Intellect Books, 1987
+
```
Comment (jrincayc) [inline tests/UnitTests-OOP.lg]: Is there any copyright issues with including this here?

File: eval.c (hunk 2 above)
Comment (brianharvey) [reply eval.c]: Depends whether @pahihu has really fixed the memory management!  :-)

File: logo.h (hunk 5 above)
Comment (brianharvey) [reply logo.h]: (Sorry I'm a little slow getting to this...)  Yes, we are no longer supporting DOS, so all the ZTC stuff can go.

File: config.h.osx (hunk 1 above)
Comment (pahihu) [reply config.h.osx]: Agreed, just pushed everything in the pull request.
Please remove it and makefile.osx as well.

File: eval.c [hunk 10]
```diff
@@ -154,7 +158,7 @@ NODE *restname, *restline;
 /* These variables are all externed in globals.h */
 
 CTRLTYPE    stopping_flag = RUN;
-char	    *logolib, *helpfiles, *csls;
+const char  *logolib, *helpfiles, *csls;
```
Comment (brianharvey) [inline eval.c]: Oh that's where the const is.  Right, these aren't constant, although they don't vary _often._  Typically they're set, if at all, in the user's startup file.

File: eval.c (hunk 2 above)
Comment (pahihu) [reply eval.c]: You could remove it, I have used it to debug how USUAL procedures are resolved.

File: eval.c (hunk 3 above)
Comment (pahihu) [reply eval.c]: In eval.c USE_GCC_DISPATCH is used 2 times: it replaces the switch + goto statement in fetch_cont: with a GNU C specific goto pointer of label.

File: init.c (hunk 4 above)
Comment (pahihu) [reply init.c]: Because wxMain.cpp declares wxMacGetXXXloc() functions returning const char*, actually they are pointers to the internal buffer of a static string. The code above is the only place where logolib is replaced with a new buffer. The code uses it otherwise as a const char*.

File: logo.h (hunk 5 above)
Comment (pahihu) [reply logo.h]: I vaguely remember that some compilers does not like the struct logo_node nodes[0] declaration. I suggest to remove the ifdef and just use struct logo_node nodes[1]

File: makefile.osx (hunk 6 above)
Comment (pahihu) [reply makefile.osx]: Agreed.

File: mem.c (hunk 7 above)
Comment (pahihu) [reply mem.c]: Yes.

File: mem.c (hunk 8 above)
Comment (pahihu) [reply mem.c]: It removes the free nodes from the oldyoungs list, traversing the list just once.

File: tests/UnitTests-OOP.lg (hunk 9 above)
Comment (pahihu) [reply tests/UnitTests-OOP.lg]: I don't know, it is an example from the above book. It contains articles written by different authors. If there is an issue, I can rewrite the code.

**Comment by pahihu** [comment]
> Is there a more editable version of evaluator.pdf? (If the program that made it can create a SVG instead of a PDF that would be preferred.)
> 
evaluator.pdf is converted from the evaluator.ps file found in the main directory.
I converted one with using convert.io website, it is not perfect. Pushed to my repo.

File: mem.c [hunk 11]
```diff
@@ -143,6 +149,16 @@ BOOLEAN addseg(void) {
...
 
+#define NILP(x)         (NIL == (x))
+
+#define GC_OPT          1
+
+#ifdef GC_OPT
+#define VALID_PTR(x)    (!NILP(x))
+#else
+#define VALID_PTR(x)    (valid_pointer(x))
+#endif
```
Comment (brianharvey) [inline mem.c]: I admit it's been a long time since I've worked on this, so maybe I'm missing something, but the reason for using valid_pointer() is that if a node is of type other than pair, the car/cdr/obj of that node may be something other than a pointer to a node, e.g., a pointer to a character string, or the address of the continuation of a special form.

File: tests/UnitTests-OOP.lg (hunk 9 above)
Comment (brianharvey) [reply tests/UnitTests-OOP.lg]: Nah, I promise Gary won't yell at us.

**Comment by brianharvey** [comment]
About evaluator.svg, when I tried printing it from Inkscape on my Mac, the ligatures in the caption at the bottom didn't work.  Is there a way to embed the font in an svg?

File: eval.c [hunk 12]
```diff
@@ -166,7 +170,7 @@ void numpush(FIXNUM obj, NODE **stack) {
     NODE *temp = newnode(CONT); /*GC*/
 
     temp->n_car = (NODE *)obj;
-    temp->n_cdr = *stack;
+    setcdr(temp, *stack); /*GC*/
```
[the remark this answers is not in the document]
Comment (pahihu) [reply eval.c]: Thanks, I see. The same construct is used in mem.c for cons(). I've removed the change.

File: mem.c (hunk 11 above)
Comment (pahihu) [reply mem.c]: I've concluded that the only valid usage for using valid_pointer() when marking the CPU stack, which can contain an arbitrary value, so we need to check them. The other locations in mem.c only checks specific types of nodes, which should contain a valid pointer in their car/cdr/obj. E.g. it won't check numbers, strings etc.

**Comment by jrincayc** [comment]
> > Is there a more editable version of evaluator.pdf? (If the program that made it can create a SVG instead of a PDF that would be preferred.)
> 
> evaluator.pdf is converted from the evaluator.ps file found in the main directory.
> I converted one with using convert.io website, it is not perfect. Pushed to my repo.

Okay, if the original is .ps at this point, .pdf is fine.

File: tests/UnitTests-OOP.lg (hunk 9 above)
Comment (jrincayc) [reply tests/UnitTests-OOP.lg]: Okay, sounds good.

**Comment by jrincayc** [comment]
@pahihu 
I would like to get this patch or a portion of it into the master branch because it fixes multiple tests. Several of the comments have not had pushes to resolve them yet.  Do you have time to work on fixing them? (If not I can create my own branch and make the edits I think it needs.)
Thanks.

**Comment by pahihu** [comment]
At the moment I do not have time to resolve the issues, but feel free to pull them in your own branch and make the edits as you like.

File: eval.c (hunk 2 above)
Comment (jrincayc) [reply eval.c]: Okay, these can be left in.

File: config.h.osx (hunk 1 above)
Comment (jrincayc) [reply config.h.osx]: Okay, removed.

File: eval.c (hunk 3 above)
Comment (jrincayc) [reply eval.c]: Removed do_label define and changed next line to ```do_list(&&x,)```

Comment (jrincayc) [reply eval.c]: Ah, I see do_label is used multiple times in the do_list macro. My mistake.

File: init.c (hunk 4 above)
Comment (jrincayc) [reply init.c]: Hm, I am going to revert this cast, but I think there is an issue, I'm just not sure of the correct fix.

File: logo.h (hunk 5 above)
Comment (jrincayc) [reply logo.h]: Removed now redundant ifdefs.

File: makefile.osx (hunk 6 above)
Comment (jrincayc) [reply makefile.osx]: Removed.

File: mem.c (hunk 7 above)
Comment (jrincayc) [reply mem.c]: Removed.

File: mem.c (hunk 8 above)
Comment (jrincayc) [reply mem.c]: Added your comment.

File: eval.c (hunk 10 above)
Comment (jrincayc) [reply eval.c]: Removed const.

**Comment by jrincayc** [comment]
I moved this over to #23

File: eval.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (jrincayc) [reply eval.c]: Fixed by @pahihu

File: mem.c (hunk 11 above)
Comment (jrincayc) [reply mem.c]: Hm, should I revert this change then? (in #23)

Comment (pahihu) [reply mem.c]: You can use the 2nd definition, which Prof. Harvey suggests.

#define VALID_PTR(x) (valid_pointer(x))

Comment (jrincayc) [reply mem.c]: Okay, I commented out ```#define GC_OPT``` but will leave the second definition in #23
